### ISSUE TITLE
Unable to resolve SQL server hostname in connection string

### ISSUE THREAD

**Issue by Ivers07** [microsoft/mssql-docker#423]
I am unable to connect to a SQL instance on a _microsoft/mssql-server-windows-developer_ container from a _microsoft/aspnet_ container.

From the aspnet container I can ping the sql server hostname successfully but if the connection string contains the hostname it returns the error:

**"A network-related or instance-specific error occurred while establishing a connection to SQL Server. The server was not found or was not accessible. Verify that the instance name is correct and that SQL Server is configured to allow remote connections. (provider: Named Pipes Provider, error: 40 - Could not open a connection to SQL Server)".**
If an IP is used it works successfully.

I have been able to re-create the issue using the following powershell:

```powershell
$db_hostname ="dbhost"
$sapassword="DockerTest1!"

docker run -d --name sqlcontainer -h $db_hostname -e ACCEPT_EULA=Y -e sa_password=$sapassword -p 1433:1433 microsoft/mssql-server-windows-developer
docker run -d --name sqlexpresscontainer -h expresshost -e ACCEPT_EULA=Y -e sa_password=$sapassword -p 1434:1433 microsoft/mssql-server-windows-express
docker run -d --name aspcontainer -h asphost microsoft/aspnet

$sqlhealth = ""
while ( $sqlhealth -ne "healthy" ) 
{
    Write-Host "waiting for SQL to be healthy"
    Start-Sleep 2
    $sqlhealth = Invoke-Expression "docker inspect -f='{{.State.Health.Status}}' sqlcontainer"
} 

$psTestcmd = @"
hostname;
ping $db_hostname;
`$SqlConnection = New-Object System.Data.SqlClient.SqlConnection;
`$SqlConnection.ConnectionString = \"Server = $db_hostname; User ID = sa; Password = $sapassword;\";
`$SqlCmd = New-Object System.Data.SqlClient.SqlCommand;
`$SqlCmd.CommandText = \"SELECT 'Connection to SQL Server successfull'\";
`$SqlCmd.Connection = `$SqlConnection;
`$SqlAdapter = New-Object System.Data.SqlClient.SqlDataAdapter;
`$SqlAdapter.SelectCommand = `$SqlCmd;
`$DataSet = New-Object System.Data.DataSet;
`$SqlAdapter.Fill(`$DataSet); 
"@

docker exec sqlcontainer powershell -command "$psTestcmd"
docker exec sqlexpresscontainer powershell -command "$psTestcmd"
docker exec aspcontainer powershell -command "$psTestcmd"
```

The SQL containers are able to resolve the DNS addresses but the ASP container cannot even though it can ping the DNS hostname.

It seems to be similar to this issues [moby/moby#31093](https://github.com/moby/moby/issues/31093) but I have check the reg values and all containers are set to 0.

Version information:

```powershell
PS C:\> docker inspect -f "{{.OsVersion}}"  microsoft/aspnet:latest
10.0.17763.316
PS C:\> docker inspect -f "{{.OsVersion}}"  microsoft/mssql-server-windows-developer:latest
10.0.14393.2007
```

```powershell
Client: Docker Engine - Community
 Version:           18.09.1
 API version:       1.39
 Go version:        go1.10.6
 Git commit:        4c52b90
 Built:             Wed Jan  9 19:34:26 2019
 OS/Arch:           windows/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          18.09.1
  API version:      1.39 (minimum version 1.24)
  Go version:       go1.10.6
  Git commit:       4c52b90
  Built:            Wed Jan  9 19:50:10 2019
  OS/Arch:          windows/amd64
  Experimental:     false
```

**Comment by forkazmoden** [comment]
Hi,

Your SQL Server container is created well. I can connect to the SQL Server instance with the hostname outside the container. 
 
Is there any special reasons for using Windows containers? With LINUX container, you can easily use an entrypoint to your aspnet image and make these two micro-services interact as independent services. Please refer to [Quickstart: Compose and ASP.NET Core with SQL Server](https://docs.docker.com/compose/aspnet-mssql-compose/).

**Comment by Ivers07** [comment]
Hi,

I am working on a proof of concept for migrating our companies current applications to containers (rather than VM's) using Docker Desktop. They are mostly ASP.net framework websites with MS SQL backend. I need to run the ASP.Net sites on windows containers so we have to run SQL on windows too. Upgrading to .Net core is not an option in the short term.

I have no issues with the creation of the SQL container, but some containers have issues accessing it via hostname in connection string. (IPs are OK so the SQL instance is up and running ok)

I have found that some parent images work successfully but other don't e.g. 
microsoft/windowsservercore:latest connects OK but
microsoft/dotnet-framework:latest does not.

**Comment by Ivers07** [comment]
I have managed to track down that the issue is due to some changed between a base image of Windows Server Core (microsoft/windowsservercore) between tags/versions ltsc2016 (1607) and 1803.
I have updated the script to re-produce the issue and here are the results i get:

### Test Script
``` Powershell
$db_hostname ="dbhost"
$sapassword="DockerTest1!"

docker run -d --name sqlcontainer -h $db_hostname -e ACCEPT_EULA=Y -e sa_password=$sapassword -p 1433:1433 microsoft/mssql-server-windows-developer

$sqlhealth = ""
while ( $sqlhealth -ne "healthy" ) 
{
    Write-Host "waiting for SQL to be healthy"
    Start-Sleep 2
    $sqlhealth = Invoke-Expression "docker inspect -f='{{.State.Health.Status}}' sqlcontainer"
} 

$psTestcmd = @"
echo ------------------------------------------------------------
hostname;
echo ------------------------------------------------------------
ping $db_hostname;
`$SqlConnection = New-Object System.Data.SqlClient.SqlConnection;
`$SqlConnection.ConnectionString = \"Server = $db_hostname; User ID = sa; Password = $sapassword;\";
`$SqlCmd = New-Object System.Data.SqlClient.SqlCommand;
`$SqlCmd.CommandText = \"SELECT 'Connection to SQL Server successfull'\";
`$SqlCmd.Connection = `$SqlConnection;
`$SqlAdapter = New-Object System.Data.SqlClient.SqlDataAdapter;
`$SqlAdapter.SelectCommand = `$SqlCmd;
`$DataSet = New-Object System.Data.DataSet;
`$SqlAdapter.Fill(`$DataSet);
`$DataSet.Tables[0];
"@

docker run -h "windowsservercoreltsc2016" microsoft/windowsservercore:ltsc2016 powershell "$psTestcmd"
docker run -h "windowsservercore1803" microsoft/windowsservercore:1803 powershell "$psTestcmd"

```

### Results:
``` Powershell
c9e139a6430c6a61036bf8ff97b6d326db656ff41ca1f933833eb6cf541da672
waiting for SQL to be healthy
waiting for SQL to be healthy
waiting for SQL to be healthy
waiting for SQL to be healthy
waiting for SQL to be healthy
waiting for SQL to be healthy
waiting for SQL to be healthy
waiting for SQL to be healthy
waiting for SQL to be healthy
waiting for SQL to be healthy
waiting for SQL to be healthy
waiting for SQL to be healthy
waiting for SQL to be healthy
waiting for SQL to be healthy
------------------------------------------------------------
windowsservercoreltsc2016
------------------------------------------------------------

Pinging dbhost [fe80::e0c5:ddc6:566c:77d7%5] with 32 bytes of data:
Reply from fe80::e0c5:ddc6:566c:77d7%5: time<1ms
Reply from fe80::e0c5:ddc6:566c:77d7%5: time=1ms
Reply from fe80::e0c5:ddc6:566c:77d7%5: time=12ms
Reply from fe80::e0c5:ddc6:566c:77d7%5: time<1ms

Ping statistics for fe80::e0c5:ddc6:566c:77d7%5:
    Packets: Sent = 4, Received = 4, Lost = 0 (0% loss),
Approximate round trip times in milli-seconds:
    Minimum = 0ms, Maximum = 12ms, Average = 3ms
1

Column1
-------
Connection to SQL Server successfull


------------------------------------------------------------
windowsservercore1803
------------------------------------------------------------

Pinging dbhost [fe80::e0c5:ddc6:566c:77d7%4] with 32 bytes of data:
Reply from fe80::e0c5:ddc6:566c:77d7%4: time=1ms
Reply from fe80::e0c5:ddc6:566c:77d7%4: time<1ms
Reply from fe80::e0c5:ddc6:566c:77d7%4: time<1ms
Reply from fe80::e0c5:ddc6:566c:77d7%4: time<1ms

Ping statistics for fe80::e0c5:ddc6:566c:77d7%4:
    Packets: Sent = 4, Received = 4, Lost = 0 (0% loss),
Approximate round trip times in milli-seconds:
    Minimum = 0ms, Maximum = 1ms, Average = 0ms
Exception calling "Fill" with "1" argument(s): "A network-related or
instance-specific error occurred while establishing a connection to SQL
Server. The server was not found or was not accessible. Verify that the
instance name is correct and that SQL Server is configured to allow remote
connections. (provider: Named Pipes Provider, error: 40 - Could not open a
connection to SQL Server)"
At line:13 char:1
+ $SqlAdapter.Fill($DataSet);
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [], MethodInvocationException
    + FullyQualifiedErrorId : SqlException
```

### WORKAROUND
I need to use the microsoft/aspnet image so i targeted that to use the 4.7.2-windowsservercore-ltsc2016 image instead of the default 4.7.2-windowsservercore-1803 that was created by visual studio and it works as required.

I've done further tests and found that there must be some for of issue introduced into WindowsServerCore between 1709 and 1803. Everything up to and including 1709 seems to work.

**Comment by selectSt4r** [comment]
LTSC2016 did not work for me as a workaround.  However, 1709 did.  I wonder what the issue is.

**Comment by fourpastmidnight** [comment]
In my case, I created my own SQL Server Windows Container using the provided Microsoft Dockerfile for Developer Edition. I customized a few things, but the one thing I _did not_ customize were the protocols that are enabled/disabled: `/TCPENABLE=1 /NPENABLE=0`.

When I attempted to connect to the SQL Server instance via an application in another container, I got the same error as the OP:

> A network-related or instance-specific error occurred while establishing a connection to SQL Server. The server was not found or was not accessible. Verify that the instance name is correct and that SQL Server is configured to allow remote connections. (provider: **Named Pipes Provider**, error: 40 - Could not open a connection to SQL Server).

Notice the provider (emphasis in the message above mine). Ah, but the installation disabled named pipes. So, why is the connection being made by named pipes?? I haven't figured out why this is. I have a few hypotheses:

1. I'm using only the hostname to connect, not a FQDN or IP Address, _and/or_ I'm not specifying a TCP port (because I'm relying on using the default port 1433 on the container network), e.g. `Server=myapp-db;`
2. The order of the protocols the SQL Server is configured to use has Named Pipes above TCP/IP?? This doesn't make much sense to me, but, it's a hypothesis.
3. I may be remembering the order of events I took incorrectly but, in my `docker-compose.yml` file, I was not specifying a `hostname:`, and I'm not using swarm mode. So, IIRC, since I'm not using swarm mode, there's no overlay network being created--therefore, no DNS server and all that good stuff. So, the containers all got a random host name. Therefore, when attempting to connect to `myapp-db`, the _client application_ attempted to connect using all available protocols, and Named Pipes happened to be the last one, and that's the one being shown in the error message. It just so happened that I added a `hostname:` to my database service in my compose file _and_ at the same time, added `tcp:` to the database server name in my connection string--and so, perhaps the `tcp:` didn't really do anything (except ensure only TCP is used for connections, as opposed to trying Shared Memory, then TCP, and then finally Named Pipes), but rather, the host was now found when TCP was the protocol being used. This is the most likely explanation. (I'll try to vet this out--and update this issue with my results.)

**Comment by fourpastmidnight** [comment]
**CONFIRMED**: Number 3 above is the correct hypothesis. I did two things at once.

So, for me, to solve the issue, I needed to make sure that all of my containers have a known, reachable hostname--because I'm not running using any orchestrator (e.g. Docker Swarm, K8s, etc.). Once the database container had a hostname, I was able to connect _without_ specifying `tcp:` in the `Data Source` or `Server` values of the connection string.
